wallet-new/src/Cardano/Wallet/Kernel/DB/Sqlite.hs

```diff
@@ -0,0 +1,562 @@
+      in formatToString ("id = " % build % ", address = " % build % ", coin = " % build)
+                        txid
+                        (_txCoinDistributionTableAddress txCoin)
+                        (_txCoinDistributionTableCoin txCoin)
+
+instance Beamable TxCoinDistributionTableT
+
+-- | The inputs' table.
+newtype TxInputT f = TxInput  {
+  _getTxInput  :: (TxCoinDistributionTableT f)
+  } deriving Generic
+
+type TxInput = TxInputT Identity
+deriving instance Show TxInput
+
+-- | Generalisation of 'mkInputs' and 'mkOutputs'.
+mkCoinDistribution :: forall a. Kernel.TxMeta
+                   -> (Getter Kernel.TxMeta (NonEmpty (Core.Address, Core.Coin)))
+                   -> (TxCoinDistributionTable -> a)
+                   -> NonEmpty a
+mkCoinDistribution txMeta getter builder =
+    let distribution  = txMeta ^. getter
+        txid          = txMeta ^. Kernel.txMetaId
+    in fmap (mk txid) distribution
+  where
+      mk :: Core.TxId -> (Core.Address, Core.Coin) -> a
+      mk tid (addr, amount) = builder (TxCoinDistributionTable addr amount (TxIdPrimKey tid))
+
+-- | Convenient constructor of a list of 'TxInput' from a 'Kernel.TxMeta'.
+mkInputs :: Kernel.TxMeta -> NonEmpty TxInput
+mkInputs txMeta = mkCoinDistribution txMeta Kernel.txMetaInputs TxInput
+
+instance Beamable TxInputT
+
+instance Table TxInputT where
+    data PrimaryKey TxInputT f = TxInputPrimKey (Beam.Columnar f Core.Address) (Beam.PrimaryKey TxMetaT f) deriving Generic
+    primaryKey (TxInput i) = TxInputPrimKey (_txCoinDistributionTableAddress i) (_txCoinDistributionTxId i)
+
+instance Beamable (PrimaryKey TxInputT)
+
+-- | The outputs' table.
+newtype TxOutputT f = TxOutput {
+  _getTxOutput :: (TxCoinDistributionTableT f)
+  } deriving Generic
+
+type TxOutput = TxOutputT Identity
+deriving instance Show TxOutput
+
+-- | Convenient constructor of a list of 'TxOutput from a 'Kernel.TxMeta'.
+mkOutputs :: Kernel.TxMeta -> NonEmpty TxOutput
+mkOutputs txMeta = mkCoinDistribution txMeta Kernel.txMetaOutputs TxOutput
+
+instance Beamable TxOutputT
+
+instance Table TxOutputT where
+    data PrimaryKey TxOutputT f = TxOutputPrimKey (Beam.Columnar f Core.Address) (Beam.PrimaryKey TxMetaT f) deriving Generic
+    primaryKey (TxOutput o) = TxOutputPrimKey (_txCoinDistributionTableAddress o) (_txCoinDistributionTxId o)
+
+instance Beamable (PrimaryKey TxOutputT)
+
+
+-- Orphans & other boilerplate
+
+instance HasSqlValueSyntax SqliteValueSyntax Core.TxId where
+    sqlValueSyntax txid = sqlValueSyntax (sformat hashHexF txid)
+
+instance HasSqlValueSyntax SqliteValueSyntax Core.Coin where
+    sqlValueSyntax = sqlValueSyntax . Core.unsafeGetCoin
+
+-- NOTE(adn) Terribly inefficient, but removes the pain of dealing with
+-- marshalling and unmarshalling of 'Integer'(s), as there is no 'HasSqlValueSyntax'
+-- defined for them.
+-- Using 'Scientific' would be a possibility, but it breaks down further as
+-- there is no 'FromField' instance for them.
```

Comment:
We could treat the `Integer` as a `Word64`, which almost certainly has a relevant instance. A `Word64` can represent POSIX timestamps up to May 28th, 586524 :smile: 

```haskell
λ> import Data.Time
λ> import Data.Time.Clock.POSIX
λ> import Data.Word
λ> :set -XNumDecimals
λ> posixSecondsToUTCTime (fromIntegral ((maxBound :: Word64) `div` 1e6))
586524-01-19 08:01:49 UTC
```

Reply:
Probably by that date planet Earth won't exist anymore anyway, so I guess that's enough precision 😅

Reply:
Be careful of not accidentally having only _seconds_ resolution. We surely want more than that, otherwise we may have to worry about rounding issues. Generally, a representation as a pair `(Int64, Word32)` is a good choice when serializing timestamps. `time` provides [`Data.Time.Clock.System.SystemTime`](https://hackage.haskell.org/package/time-1.9.1/docs/Data-Time-Clock-System.html#v:SystemTime) having that representation, which is exactly what's provided by system calls returning timestamps.

Reply:
The `Timestamp` is stored using `Microseconds` resolution, and that's what we're storing in the database. An integral number of microseconds since the epoch